kroxylicious/src/main/java/io/kroxylicious/proxy/KafkaProxy.java

```diff
@@ -89,31 +90,44 @@ public KafkaProxy startup() throws InterruptedException {
             throw new IllegalStateException("This proxy is already running");
         }
 
+        var portConflictDefector = new PortConflictDetector();
+        portConflictDefector.validate(virtualClusterMap);
+
         var availableCores = Runtime.getRuntime().availableProcessors();
         var meterRegistries = new MeterRegistries(micrometerConfig);
 
-        this.adminEventGroup = buildNettyEventGroups(availableCores, config.isUseIoUring());
-        this.serverEventGroup = buildNettyEventGroups(availableCores, config.isUseIoUring());
+        this.adminEventGroup = buildNettyEventGroups("admin", availableCores, config.isUseIoUring());
+        this.serverEventGroup = buildNettyEventGroups("server", availableCores, config.isUseIoUring());
 
         maybeStartMetricsListener(adminEventGroup, meterRegistries);
 
-        this.endpointProviders = virtualClusterMap.entrySet().stream()
-                .collect(Collectors.toMap(e -> e.getValue().getClusterEndpointProvider(), Map.Entry::getValue));
+        var tlsServerBootstrap = buildServerBootstrap(serverEventGroup, new KafkaProxyInitializer(config, true, endpointRegistry, false, Map.of()));
+        var plainServerBootstrap = buildServerBootstrap(serverEventGroup, new KafkaProxyInitializer(config, false, endpointRegistry, false, Map.of()));
 
-        var networkBindings = buildNetworkBindings();
+        var unused = networkBindingExecutor.submit(() -> {
```

Comment:
Why drive the `take -> select bootstrap -> perform` out here? It seems like quite binding specific work for the KafkaProxy to do.

Could we instead install the bootstraps into the registry and it could be responsible for the executor/loop itself? Alternatively if we want to drive the work externally, perhaps it could be refactored like `registry.drive(tlsServerBoorstrap, plainServerBootstrap)`.

Reply:
Yes, that's a good thought. When I started off I was trying to keep the `EndpointRegistry` free of knowledge of Netty. This division of responsibility is a reflection of that.  However the way things developed eroded that idea -`EndpointRegistry` now knows about Netty Channels in its data structures and uses Netty Channel Attribute.

I think the case for the separation of binding responsibility is no longer valid.

Reply:
We can possibly do both things by injecting a wrapper into the `EndpointRegistry` which takes the bind requests and translates them into netty specific actions (though I'm not sure its needed to keep netty out of the picture)